containers/ws: Support host-specific ssh keys

benniekiss opened the pull request:
This PR adds the ability to set host-specific ssh keys via env-var in the cockpit-ws container.

Users can pass environment variables in the format of `COCKPIT_SSH_KEY_PATH_{HOSTNAME}`, where `HOSTNAME` is the hostname used in the `Connect to` field of the cockpit login page, and cockpit will use the configured key to login to the host. If no host-specific key is set, it falls back to using the `COCKPIT_SSH_KEY_PATH` environment variable.

I still need to make a unit test, but I wanted to ask for suggestions as to the best route. The [cockpit bastion test](https://github.com/cockpit-project/cockpit/blob/f071d8cd2a20b309dae72392a7a49223d2aaa3c1/test/verify/check-ws-bastion#L162-L219) could be replicated, or it could be refactored to test host-specific keys. I wasn't sure what the best approach would be, and would appreciate any advice before proceeding!

This functionality will enable users to login to multiple hosts via ssh keys once the host-switcher is deprecated, providing a solution to user concerns such as  #20901

## cockpit/ws container: Support host specific SSH keys

The [cockpit/ws](https://quay.io/repository/cockpit/ws) container now supports adding a multiple SSH private keys. In addition to the existing `$COCKPIT_SSH_KEY_PATH` environment variable you can now set host specific `$COCKPIT_SSH_KEY_PATH_`*HOSTNAME* variables, where `HOSTNAME` is the host name (in capital letters) used in the `Connect to:` field of the login page.

_Thanks to [benniekiss](https://github.com/benniekiss) for contributing this feature!_

martinpitt commented:
Thanks @benniekiss ! This is a pretty cool feature! It does need an integration test, yes. I suggest you factorize that test a bit, with a `do_test_key_login(options: str)`, and then call this several times in `testKeyLogin()` with the scenarios (only COCKPIT_SSH_KEY_PATH, the per-hostname ones, and the fallback). You can use invalid keys for the ones that you don't expect to get picked of course, so you don't need multiple VMs/SSH targets.

benniekiss commented:
I refactored the test to account for setting a host-specific key var and the fallback. I was not able to run these tests locally. I was having a lot of trouble getting the test infrastructure up on my aarch64 system -- M2 macbook on asahi. I tried following the docs in HACKING.md and in the test/README.md, but I struggled to get it working.

martinpitt commented on containers/ws/README.md:
```diff
@@ -97,11 +97,15 @@ can mount your known host keys into the container at
+    -v ~/.ssh/:/.ssh:ro,Z
 
-Then cockpit will use the provided password to decrypt the key and establish an SSH connection to the given host using that private key.
+Private keys must be configured with a password so that cockpit can use the provided password to decrypt the key and establish an SSH connection to the given host.
```
That's not a requirement -- passwordless keys are supported (then the password is unused and can be empty). How about

> Private keys can be encrypted; then cockpit can use the provided password to decrypt the key.

(The "establish connection.." is fairly redundant, so how about we just drop it?)

benniekiss replied:
Oh, I didn't know that about passwordless keys. This sounds good to me!

martinpitt reviewed (changes requested):
Cheers! This looks pretty good, I'd just like to adjust the documentation. Please also squash the commits together into a "containers/ws: Support host-specific ssh keys".

I triggered the fedora-coreos tests (the only relevant ones) to see how it fares.

martinpitt commented:
The test [failed in the second round](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20904-85867373-20240926-121250-fedora-coreos-other/log.html#159), presumably as it tries to regenerate the RSA key a second time when it already exists. This needs to be taken out of the loop.

martinpitt commented:
@benniekiss as previously: Please tell me when you are fed up and want me to take over :grin:

benniekiss commented:
rebased on main and squashed! 

I moved the `ssh-keygen` calls and the `self.addCleanup` calls outside of the loop, and I updated the documentation as suggested.

If the tests don't pass, I'd appreciate if you'd be able to takeover and see this through! Just because I still haven't been able to get the tests running locally

benniekiss commented:
fixed a misplaced comment

martinpitt commented:
@benniekiss Still unhappy.. I'll take over, thank you!

benniekiss commented:
You're welcome! And thank you, too

martinpitt commented:
[Fixed](https://github.com/cockpit-project/cockpit/compare/53611e494ba129e63c71a1e962710e54df0f63b2..1ca735c9616a971592f60a729add1997c90ab5fd) ; it was missing the .upper() for the hostname and some cleanup.

Thanks!

martinpitt commented:
@benniekiss I added a release note to the description. Proof-reading much appreciated!

benniekiss commented:
Release note looks good overall! I cleaned up some grammar. 

This markdown formatting is a little hard to read, but I wasn't sure if there was a clearer way to write it.

> `$COCKPIT_SSH_KEY_PATH_`*HOSTNAME*

I had used `COCKPIT_SSH_KEY_PATH_{HOSTNAME}` previously, so that would be my suggestion.

martinpitt commented:
@benniekiss you are right, I changed it to {}. Cheers!
